templates/ikfast61_moveit_plugin_template.cpp

```diff
@@ -958,6 +1044,202 @@ bool IKFastKinematicsPlugin::getPositionIK(const geometry_msgs::Pose &ik_pose,
+      {
+        result.kinematic_error = kinematics::KinematicErrors::IK_SEED_OUTSIDE_LIMITS;
+        ROS_ERROR_STREAM("ik seed is out of bounds");
+        return false;
+      }
+
+    }
+
+
+    // computing all solutions sets for each sampled value of the redundant joint
+    if(!sampleRedundantJoint(options.discretization_method,sampled_joint_vals))
+    {
+      result.kinematic_error = kinematics::KinematicErrors::UNSUPORTED_DISCRETIZATION_REQUESTED;
+      return false;
+    }
+
+    for(unsigned int i = 0; i < sampled_joint_vals.size(); i++)
+    {
+      vfree.clear();
+      vfree.push_back(sampled_joint_vals[i]);
+      numsol += solve(frame,vfree,ik_solutions);
+      solution_set.push_back(ik_solutions);
+    }
+  }
+  else
+  {
+    // computing for single solution set
+    numsol = solve(frame,vfree,ik_solutions);
+    solution_set.push_back(ik_solutions);
+  }
+
+  ROS_DEBUG_STREAM_NAMED("ikfast","Found " << numsol << " solutions from IKFast");
+  bool solutions_found = false;
+  std::stringstream ss;
+  if( numsol > 0 )
+  {
+    for(unsigned int r = 0; r < solution_set.size() ; r++)
+    {
+
+      ik_solutions = solution_set[r];
+      numsol = ik_solutions.GetNumSolutions();
+      for(int s = 0; s < numsol; ++s)
+      {
+        std::vector<double> sol;
+        getSolution(ik_solutions,s,sol);
+        ss.str("");
+        ss<<"[";
+        for(unsigned int i = 0 ; i < sol.size() ; i++)
+        {
+          ss<<sol[i]<<" ";
+        }
+        ss<<"]";
+        ROS_DEBUG_NAMED("ikfast","Sol %d: %s", s, ss.str().c_str());
+
+        bool obeys_limits = true;
+        for(unsigned int i = 0; i < sol.size(); i++)
+        {
+          // Add tolerance to limit check
+          if(joint_has_limits_vector_[i] && ( (sol[i] < (joint_min_vector_[i]-LIMIT_TOLERANCE)) ||
+                                              (sol[i] > (joint_max_vector_[i]+LIMIT_TOLERANCE)) ) )
+          {
+            // One element of solution is not within limits
+            obeys_limits = false;
+            ROS_DEBUG_STREAM_NAMED("ikfast","Not in limits! " << i << " value " << sol[i] << " has limit: " << joint_has_limits_vector_[i] << "  being  " << joint_min_vector_[i] << " to " << joint_max_vector_[i]);
+            break;
+          }
+        }
+        if(obeys_limits)
+        {
+          // All elements of solution obey limits
+          solutions_found = true;
+          solutions.push_back(sol);
+        }
+      }
+    }
+
+    if(solutions_found)
+    {
+      result.kinematic_error = kinematics::KinematicErrors::OK;
+      return true;
+    }
+  }
+  else
+  {
+    ROS_DEBUG_STREAM_NAMED("ikfast","No IK solution");
+  }
+
+  result.kinematic_error = kinematics::KinematicErrors::NO_SOLUTION;
+  return false;
+}
+
+bool IKFastKinematicsPlugin::sampleRedundantJoint(kinematics::DiscretizationMethod method, std::vector<double>& sampled_joint_vals) const
```

Comment:
Is there only ever ONE redundant joint (this may be a requirement for IKFast, I don't know)?